File: frontend/angular-personal-finance-tracker/src/app/common/budget-with-spending.ts
// budget-with-spending.ts (create in src/app/common/)
export interface BudgetWithSpending {
  // Original budget fields
  id?: number;
  categoryId: number;
  categoryName?: string; // For display
  budgetAmount: number;
  userId: number;
  
  startDate?: Date;
  endDate?: Date;
  
  // Spending calculation fields
  spent: number;        // Total spent in this category during period
  percentage: number;      // (spending / amount) * 100
  remaining: number;       // amount - spending
  isOverBudget: boolean;   // percentage > 100
  nearThreshold: boolean; // percentage >= alertThreshold (default 80)
  alertThreshold?: number; // Default 80%
}

File: backend/personal-finance-tracker-spring-boot/src/main/java/com/petruth/personal_finance_tracker/dto/BudgetWithSpending.java
package com.petruth.personal_finance_tracker.dto;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * Response DTO that includes budget + calculated spending data
 */
public record BudgetWithSpending(
        Long id,
        String categoryName,
        Long categoryId,
        BigDecimal budgetAmount,
        BigDecimal spent,
        BigDecimal remaining,
        Double percentage,
        Boolean isOverBudget,
        Boolean nearThreshold,
        Integer alertThreshold,
        LocalDate startDate,
        LocalDate endDate
) {
    // Helper method to check if alert should be sent
    public boolean shouldSendAlert() {
        return nearThreshold || isOverBudget;
    }

    public String getAlertType() {
        if (isOverBudget) {
            return "OVER_BUDGET";
        } else if (nearThreshold) {
            return "WARNING";
        }
        return "NONE";
    }
}